sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala

```diff
@@ -102,6 +103,14 @@ case class AdaptiveSparkPlanExec(
     OptimizeLocalShuffleReader(conf)
   )
 
+  @transient private val finalStageOptimizerRules: Seq[Rule[SparkPlan]] =
+    context.qe.sparkPlan match {
+      case _: DataWritingCommandExec =>
```

Comment:
we need to match all writing commands, including DS v1, v2 and file source. Maybe we can create a tagging trait like `UserDefinedExpression`, to tag these writing commands.

Reply:
It seems DSv2 is not ready for write as per https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala#L988-L994.

Meanwhile, will it too big a change for those interfaces to extend the tagging trait ?

Reply:
File source v2 is not ready yet, but it doesn't mean DS v2 is not ready for writing. Please follow `InsertAdaptiveSparkPlan.applyInternal`

Reply:
Sure, is there a UT for DS v2 write ? I find only V1 is used for write no matter format I specify.

Reply:
See `DataSourceV2Suite`, we have testing v2 source that support writing.

Reply:
I've added match for `V2TableWriteExec` and UT. Not sure I've got it right as v2 API is quite new to me. Please help review. Thanks.